Makefile

```diff
@@ -691,7 +703,8 @@ get_pkg_build_dev_yml = $(if $(wildcard pkg/$1/build-dev.yml),build-dev.yml,buil
 
 eve-%: pkg/%/Dockerfile build-tools $(RESCAN_DEPS)
 	$(QUIET): "$@: Begin: LINUXKIT_PKG_TARGET=$(LINUXKIT_PKG_TARGET)"
-	$(QUIET)$(LINUXKIT) $(DASH_V) pkg $(LINUXKIT_PKG_TARGET) $(LINUXKIT_OPTS) -build-yml $(call get_pkg_build_yml,$*) pkg/$*
+	$(eval LINUXKIT_DOCKER_LOAD := $(if $(filter $(PKGS_DOCKER_LOAD),$*),--docker,))
+	$(QUIET)$(LINUXKIT) $(DASH_V) pkg $(LINUXKIT_PKG_TARGET) $(LINUXKIT_OPTS) $(LINUXKIT_DOCKER_LOAD) --platforms linux/$(ZARCH) -build-yml $(call get_pkg_build_yml,$*) pkg/$*
```

Comment:
Can we also tag the image without arch suffix here in case of `HOSTARCH` not equals with `ZARCH` (`CROSS` equals with 1)? Right now it is required by scripts in tools to have images without suffix in local docker cache.

Reply:
I don't understand. What this is doing is, if the package is on the "load into docker" list, then:

```
lkt pkg build --docker pkg/foo
```

That should load:

* `lfedge/eve-foo:abcd-amd64` for amd64
* `lfedge/eve-foo:abcd-arm64` for arm64 (etc)
* `lfedge/eve-foo:abcd` for whatever arch we currently are on

What else do you want to do?

Reply:
Is that case even useful?

If I do

```
docker run alpine:3.16
```

it will resolve to my local platform. If the index for `alpine:3.16` does not exist for my local platform, I will get an error.

When is it useful, such that we use it, that have `lfedge/eve-foo:abcd` for something other than the platform on which we run is useful?

Reply:
> Agree, it is not the best approach. But it is what we have in scripts in tools right now

That was my point. I don't think it is. Even in scripts, when we run:

```
docker run lfedge/eve:abcd
```

that _always_ automatically resolves to "`abcd` for my platform". Unless we explicitly request it (and have qemu set up, etc.), it never will resolve to "`abcd` for some other platform". If `abcd` doesn't exist for the platform on which I am running, it is an error.

So when will we ever need `abcd` to point to anything other than "my platform (i.e. platform on which I am running"?

Reply:
And I agree 200% that if we need to change scripts, let's change them in a different PR after this is done.

Reply:
Taking that apart:

1. I made changes in pkg/mkconf - ✅ 
2. I run something like `make ZARCH=riscv64 pkgs` - this then builds and loads into docker `lfedge/eve-mkconf:abcd-riscv64` ✅ 
3. I want to run `make ZARCH=riscv64 HV=mini eve` - this eventually calls `makeconfig.sh`, which runs `docker run ... ${MKCONFIG_TAG}`, which resolves to `docker run lfedge/mkconf:abcd` ❌ 

Agreed. That fails. Because it assumes it is going to run `abcd` for my arch, and just happen to spit out for a different arch.

But wouldn't that fail right now, even before this PR? This PR doesn't cause it to fail any worse, does it?

Reply:
> But wouldn't that fail right now, even before this PR? This PR doesn't cause it to fail any worse, does it?

No. I can see something like this in the end of the build for riscv64 on amd64 host platform:
```
Successfully built 27ac4a3437e0
Successfully tagged lfedge/eve-mkconf:911f739be21348eb2ca0a19b944633549d76ed11-riscv64
Tagging lfedge/eve-mkconf:911f739be21348eb2ca0a19b944633549d76ed11-riscv64 as lfedge/eve-mkconf:911f739be21348eb2ca0a19b944633549d76ed11
```

Reply:
Is that from a CI run that you can link to? Or did you run that locally? If so, what command did you run?

Reply:
From local commands (we have such commands in publish workflow, but there we do push):
```
rm build-tools/bin/linuxkit
make LINUXKIT_VERSION=master LINUXKIT_SOURCE=https://github.com/rvs/linuxkit.git linuxkit
LK_BUILD_ARGS='{"EVE_BUILDER_IMAGE":"lfedge/eve-alpine:81e6520f4f1554789eb1ac299168e72ac37d88e2-riscv64"}' ZARCH=riscv64 make pkg/mkconf
```
_do not forget to remove hacked linuxkit after trial `rm build-tools/bin/linuxkit`_

Reply:
OK, yes, I see it.

Reply:
There's something not right about this. Tagging an architecture with its name, whether your local or a different one, is fine. Tagging the default - which should point to an index, and does when it is on registry, but not right now in docker local cache - with something other than your local architecture is wrong. It is just asking for problems.

As far as I can tell, we are doing three things here:

1. Building for a different architecture by adding the hacked build-args. We don't need that. Using `lkt pkg build --platforms linux/riscv64` will provide the exact same result, by using buildkit's native cross-build support.
2. Loading that cross-built image as `image:abcd-riscv64` into docker cache. We don't need that. Using `--docker` provides the exact same result.
3. Tagging `abcd-riscv64` as `abcd` so we can just run `abcd`, even if our platform is not riscv64. **This part is wrong**

We should not be tagging the non-local-arch version with the default tag. We should just run the local version, if that is what we want.

You are saying that if we do not do it, then, unless we ran `make pkg/mkconf` for our local arch _in addition to_ riscv64 (or some other arch), we _never_ end up with just `abcd`, and then we have nothing to `docker run`.

On the other hand, the archless tag works great when running it not from local cache (i.e. from registry), because then it resolves the index correctly.

I do **not** think making the scripts run with `-<arch>` is right, because then you are bypassing good index resolution from registries.

I am kind of stuck. What do we do to make it work in all scenarios - from remote registry and from local build - without breaking proper design? Should we make a local pkg build a prerequisite?

_This will be so much easier when they finalize that docker desktop full index support._

Reply:
This is where we are sort of stuck @giggsoff

Reply:
> I do not think making the scripts run with -<arch> is right, because then you are bypassing good index resolution from registries.

Probably we should discuss this point (as the one of the possible solution). Should we rely here on index resolution? I mean, if we plan to use cross-platform builds (agree, it is possible right now, but not sure that we support this future completely), do we expect that we will run images from another arch (from system arch) when we build image for another arch (ZARCH not equal HOSTARCH)? Will it potentially ends with some errors if we will put arch-specific binaries/logic inside of images we call from scripts? It looks like we should define arch we want to run image from scripts somewhere. If we will go this way, why not to use suffix for this?

Reply:
index resolution is the "right way" ™️ for doing these things. Whenever we do `docker run`, we are asking to run whatever is our local arch by default.

If we especially want to run something other than local (i.e. using emulation), then the "right way" ™️ is via `docker run --platform linux/other-arch image`.

I have nothing against storing an arch-specific image in a registry or a cache with a specific tag, so we can get access to it, but the official OCI and docker way of doing it with indexes is archless tags and index lookup.

The question I have is, do we _ever_ want to run something else?

* When `makeconfig.sh` runs, it is just manipulating input bits to create a config partition that has nothing arch-specific, so we should run our local (don't emulate unless we need to).
* When  `makerootfs.sh` runs, it is just is manipulating input bits from a `linuxkit build` to create an output filesystem (extr4 or squash). It isn't doing anything arch-specific, so we should run our local (don't emulate unless we need to).
* When `makeiso.sh` runs, it is just manipulating bits to form an iso, nothing arch-specific about how it manipulates.
* Same for `makeflash.sh`

In other words, not one of the ones we are loading into docker for `docker run` has to be run in emulation mode. We can build for riscv64 on amd64, and the `amd64` versions of `make*.sh` will do the job just fine (and a lot more quickly than the emulated arm64 or riscv64 versions). These are just utilities for manipulating bits.

Given that, I think we should indeed just be running our local arch version, which means use the index to default to local.

if I am building eve for `ZARCH=riscv64`, and I am on `amd64`, then I should:

1. Build packages that need to go into the eve image for riscv64
2. Build the rootfs _for_ riscv64
3. Using all of my utilities (`docker run` and scripts) for my local arch, i.e. amd64

Reply:
Will it make sense to build packages required for scripts for HOSTARCH as well as for ZARCH in that case (if they are different)?

Reply:
Yes, I think so. For packages for scripts, they are meant to be run locally, so they should be a prerequisite for scripts.

None of this matters if you haven't changed `pkg/mk*`, because then they will be in the registry with an index, and so they will just work.

If I did change it, then platforms should _also_ include the HOSTARCH, so that we get it for local, and the default tag as well.

I think that is the easiest approach. Should have something out in a few minutes.

Reply:
@giggsoff updated. Anytime we use `lkt pkg build` with `--docker`, it will ensure that it builds for `HOSTARCH` as well. It ensures the list is unique (so `make eve ZARCH=amd64` on amd64 doesn't do `--platforms linux/amd64,linux/amd64`). Since `lkt pkg build` uses `buildkit`, with its efficient caching, it should be very efficient.

Take a look. If this works, we will be ready for final review from Erik.

Reply:
It works, thank you!

Reply:
Party time!